app/src/forms/form/externalApi/routes.js

```diff
@@ -0,0 +1,36 @@
+const routes = require('express').Router();
+const { currentUser, hasFormPermissions } = require('../../auth/middleware/userAccess');
+const validateParameter = require('../../common/middleware/validateParameter');
+const P = require('../../common/constants').Permissions;
+
+const controller = require('./controller');
+
+routes.use(currentUser);
+
+routes.param('formId', validateParameter.validateFormId);
+
+routes.get('/:formId/externalAPIs', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.listExternalAPIs(req, res, next);
+});
+
+routes.post('/:formId/externalAPIs', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.createExternalAPI(req, res, next);
+});
+
+routes.get('/:formId/externalAPIs/algorithms', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.listExternalAPIAlgorithms(req, res, next);
+});
+
+routes.get('/:formId/externalAPIs/statusCodes', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.listExternalAPIStatusCodes(req, res, next);
+});
+
+routes.put('/:formId/externalAPIs/:externalAPIId', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.updateExternalAPI(req, res, next);
+});
+
+routes.delete('/:formId/externalAPIs/:externalAPIId', hasFormPermissions([P.FORM_READ, P.FORM_UPDATE]), async (req, res, next) => {
+  await controller.deleteExternalAPI(req, res, next);
+});
```

Comment:
We've previously had problems when there are two IDs in the parameters. I don't think anything is checking that the `externalAPIId` belongs to the `formId`. The exploit would be to set up a form and have FORM_READ and FORM_UPDATE on it, but then update/delete an `externalAPIId` for another form.

As an example, in the `validateParameter` middleware there are some validators like `validateFormVersionId` that both validate that the parameter is a UUID but also that it matches the `formId`.

Reply:
yes, relying on the validateFormId to do it's job, then the service will do the check for formId + externalApiID to make sure they are paired.

Reply:
I think the idea of the route / controller / service design is to do have request validation in the route (middleware) layer. Can you create another ticket for this?

Reply:
Sure. Will add a new middleware for external id